src/client/chat.js

```diff
@@ -0,0 +1,538 @@
+      const hashable = crypto.createHash('sha256').update(concat('i64', packet.salt, 'i64', packet.timestamp / 1000n, 'pstring', packet.plainMessage, 'i8', 70))
+      if (packet.formattedMessage) hashable.update(Buffer.from(packet.formattedMessage, 'utf8'))
+      for (const previousMessage of packet.previousMessages) {
+        hashable.update(concat('i8', 70, 'UUID', previousMessage.messageSender))
+        hashable.update(Buffer.from(previousMessage.messageSignature))
+      }
+      const hash = hashable.digest()
+      const verifier = crypto.createVerify('RSA-SHA256')
+      if (packet.messageSignature) verifier.update(Buffer.from(packet.messageSignature))
+      verifier.update(concat('UUID', packet.senderUuid, 'buffer', hash))
+      return verifier.verify(pubKey, Buffer.from(packet.headerSignature))
+    } else if (mcData.supportFeature('sessionSignature')) { // 1.19.3
+      const length = Buffer.byteLength(packet.plainMessage, 'utf8')
+      const previousMessages = packet.previousMessages.length > 0 ? ['i32', packet.previousMessages.length, 'buffer', Buffer.concat(...packet.previousMessages.map(msg => msg.signature))] : ['i32', 0]
+
+      const signable = concat('i32', 1, 'UUID', packet.senderUuid, 'UUID', session, 'i32', packet.index, 'i64', packet.salt, 'i64', packet.timestamp / 1000n, 'i32', length, 'pstring', packet.plainMessage, ...previousMessages)
+      return crypto.verify('RSA-SHA256', signable, pubKey, packet.messageSignature)
+    } else { // 1.19
+      const signable = concat('i64', packet.salt, 'UUID', packet.senderUuid,
+        'i64', packet.timestamp / 1000n, 'pstring', packet.signedChatContent)
+      return crypto.verify('RSA-SHA256', signable, pubKey, packet.signature)
+    }
+  }
+  client.refreshAcknowledgements = () => {
+    if (!client.chat_log) return []
+
+    client.chat_log.untracked = 0
+    return client.chat_log.acknowledgements
+  }
+
+  client.sendChatMessage = async message => {
+    if (message.length > 256) throw Error('Chat message length cannot exceed 256 characters')
+
+    if (mcData.supportFeature('signedChat')) {
+      if (mcData.supportFeature('chainedSignature')) {
+        const salt = 0n
+        const packetData = {
+          message,
+          salt
+        }
+        if (client.useChatPreview) {
+          try {
+            packetData.signedPreview = true
+            const preview = await fetchPreview(message)
+            const acknowledgements = client.refreshAcknowledgements()
+            const timestamp = BigInt(Date.now())
+            packetData.timestamp = timestamp
+            packetData.signature = client.signMessage(message, timestamp, salt, acknowledgements, false, preview)
+            packetData.previousMessages = client.chat_log ? client.chat_log.acknowledgements : []
+            packetData.lastMessage = client.chat_log ? client.chat_log.lastUntracked : undefined
+          } catch {
+            // Discard message
```

Comment:
Should not silently fail

Reply:
What do you propose instead? An event that identifies that a message has been discarded?

Reply:
Throw an exception